find_exponential/src/main.rs

```diff
@@ -0,0 +1,309 @@
+        .into_iter()
+        .filter_map(|e| e.ok())
+        .filter(|e| e.file_type().is_file())
+        .filter(|e| if let Some(ext) = e.path().extension() { ext == "rs" } else { false });
+
+    let mut literals = HashSet::new();
+    for file in walkdir {
+        if file.path() == Path::new("../src/entities.rs") {
+            continue;
+        }
+        println!("parsing {}", file.path().display());
+        let content = fs::read_to_string(file.path())
+            .expect(&format!("unable to read file {:?}", file.path()));
+        let parsed = syn::parse_file(&content)
+            .expect(&format!("unable to parse file {:?}", file.path()));
+        literals::extract_literals_from_items(&mut literals, parsed.items);
+    }
+
+    let literals: Vec<_> = literals.into_iter()
+        .filter(|lit| !lit.contains(&b'\n'))
+        .filter(|lit| !lit.is_empty())
+        .collect();
+
+    let num_batches_finished = AtomicU64::new(0);
+    let num_batches_finished = &num_batches_finished;
+    let mut rng = Xoshiro256Plus::from_rng(&mut rand::thread_rng()).unwrap();
+    let start_time = Instant::now();
+
+    // start threads
+    thread::scope(|s| {
+        let threads: Vec<_> = (0..num_cpus).map(|_| {
+            let literals = literals.clone();
+            // Get unique RNG for each thread. Read docs of Xoshiro256Plus for further information.
+            rng.jump();
+            let rng = rng.clone();
+            s.spawn(move |_| worker_thread_fn(literals, &num_batches_finished, &start_time, rng))
+        }).collect();
+
+        for thread in threads {
+            println!("Joined thread: {:?}", thread.join());
+        }
+    }).unwrap();
+}
+
+fn worker_thread_fn(
+    mut literals: Vec<Vec<u8>>, num_batches_finished: &AtomicU64, start_time: &Instant,
+    mut rng: Xoshiro256Plus,
+) {
+    loop {
+        literals.shuffle(&mut rng);
+        let combs = literals.iter()
+            .combinations(COMBINATIONS)
+            .take(BATCH_SIZE);
+        for comb in combs {
+            let concatenated = comb.into_iter().flatten().cloned().collect::<Vec<_>>();
+            // we mustn't panic here, because that would result in the thread being killed
+            let pattern = match String::from_utf8(concatenated) {
+                Ok(pattern) => pattern,
+                Err(err) => {
+                    println!("Couldn't convert pattern to UTF-8. Err: {:?}.", err);
+                    continue;
+                }
+            };
+            let res = panic::catch_unwind(|| {
+                test(&pattern);
+            });
+            if res.is_err() {
+                println!("Panic caught. Pattern: {:?}", pattern);
+            }
+        }
+
+        // print throughput
+
+        // TODO: I don't know which ordering we actually need, SeqCst is always correct, but may
+        // not actually be needed.
+        let batches_finished = num_batches_finished.fetch_add(1, Ordering::SeqCst);
```

Comment:
Relaxed ordering should be fine when you're just counting stuff.

Reply:
According to [the nomicon Relaxed section](https://doc.rust-lang.org/nomicon/atomics.html#relaxed) you're right. I thought so as well, but wasn't 100% sure.

> For instance, incrementing a counter can be safely done by multiple threads using a relaxed `fetch_add` if you're not using the counter to synchronize any other accesses.